interfaces/builtin/greengrass_support.go

```diff
@@ -126,31 +230,58 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
 mount options=(rw, bind) /proc/sys/ -> /proc/sys/,
 mount options=(rw, bind) /proc/sysrq-trigger -> /proc/sysrq-trigger,
 
-# remap a few /proc accesses to /dev/null
+# mount some devices using /dev/null
 mount options=(rw, bind) /dev/null -> /proc/kcore,
 mount options=(rw, bind) /dev/null -> /proc/sched_debug,
 mount options=(rw, bind) /dev/null -> /proc/timer_stats,
 
-# perform the pivot_root into the overlay
-pivot_root oldroot=/var/snap/greengrass/@{SNAP_REVISION}/rootfs/.pivot_root*/ /var/snap/greengrass/*/rootfs/,
-mount options=(rw, rprivate) -> /.pivot_root*/,
-umount /.pivot_root*/,
-owner /.pivot_root*/ w,
-mount options=(rw, rprivate) -> /,
-mount options=(ro, remount, rbind) -> /,
+# umounts for tearing down containers
+umount /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/**,
+
+# this is for container device creation
+# also need mknod and mknodat in seccomp
+capability mknod,
+
+# for the greengrassd pid file
+owner /{var/,}run/greengrassd.pid rw,
 
-# allow tearing down the overlay
-umount /var/snap/@{SNAP_NAME}/**,
-/run/mount/utab rw,
-/bin/umount ixr,
+# all of the rest of the accesses are made by child containers and as such are 
+# "post-pivot_root", meaning that they aren't accessing these files on the 
+# host root filesystem, but rather somewhere inside $SNAP_DATA/rootfs/
+# Note: eventually greengrass will gain the ability to specify child profiles
+# for it's containers and include these rules in that profile so they won't
+# be here, but that work isn't done yet
+# Additionally see LP bug #1791711 for apparmor resolving file accesses after
+# a pivot_root
 
-# For lambda functions, post pivot_root lambda execution accesses
+# for child container lambda certificates
 /certs/ r,
 /certs/** r,
 /group/ r,
 /group/** r,
 /state/ r,
-/state/sqlite* rwk,
+/state/{,**} r,
+# the child containers need to use a file lock here
+owner /state/secretsmanager/secrets.db krw,
+owner /state/secretsmanager/secrets.db-journal rw,
+owner /state/shadow/ krw,
```

Comment:
Do you really need 'k' for this directory?

Reply:
After looking more closely it doesn't need to lock the directory here, but the files underneath yes.